url61db808cfe_ctriposs_bigcache_tgz-pJ8-com_ctriposs_bigcache_BigCacheStressTestJ8/src/com/ctriposs/bigcache/storage/MemoryMappedStorage.java

```diff
@@ -0,0 +1,78 @@
+package com.ctriposs.bigcache.storage;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.RandomAccessFile;
+import java.nio.ByteBuffer;
+import java.nio.MappedByteBuffer;
+import java.nio.channels.FileChannel;
+
+public class MemoryMappedStorage implements IStorage {
+
+	private RandomAccessFile raf;
+	private ThreadLocalByteBuffer threadLocalBuffer;
+
+	public MemoryMappedStorage(String dir, int index, int capacity) throws IOException {
+		File backFile = new File(dir);
+		if (!backFile.exists()) {
+			backFile.mkdirs();
+		}
+		String backFileName = dir + index + "-" + System.currentTimeMillis() + DATA_FILE_SUFFIX;
+		raf = new RandomAccessFile(backFileName, "rw");
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code assigns a resource that might leak to a class field, causing a resource leak that slows down or crashes your system. For example, when FileChannel.map throws an exception, the *MemoryMappedStorage* object isn’t constructed so you can’t close this resource afterward.

**Fix**
Ensure that if an exception is thrown, the *raf* resource is closed before the constructor is exited.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).